discord_bot/discord_bot.py

```diff
@@ -1,60 +1,85 @@
+        self.websocket = None
 
-        # actions below
-        asyncio.get_event_loop().run_until_complete(self.gateway_handler())
+        # actions
+        self.event_loop.run_until_complete(self.gateway_handler())
 
     @staticmethod
     def load_config():
+        """
+        Loads the configurations
+        :return: dict - the configurations
+        """
         with open(CONFIG_FILE_PATH) as f:
             return json.load(f)
 
-    @staticmethod
-    def get_gateway():
-        r = requests.get(DISCORD_GATEWAY_ENDPOINT)
+    def get_gateway(self):
+        """
+        Caches a gateway value and retrieves a new URL
+        :return: gateway URL
+        """
+        r = requests.get(self.config("discord_gateway_endpoint"))
         return r.json()['url']
 
-    async def handshake(self, websocket):
-        identity_obj = {
-            "token": self.config["bot_token"],
-            "properties": {
-                "$os": "windows",
-                "$browser": "disco",
-                "$device": "disco"
-            },
-            "compress": False,
-            "large_threshold": 250,
-            "shard": [1, 10]
-        }
-        await websocket.send(json.dumps({"op": 2, "d": identity_obj}))
+    async def send_json(self, payload):
+        asyncio.ensure_future(self.websocket.send(json.dumps(payload)))
 
-    async def gateway_handler(self):
+    async def handshake(self, message):
+        """
+        When a websocket connection is opened, Hello payload is received. Then, an Identify payload is sent as
+        part of the handshake to authorize this client.
 
+        :param message: dict - hello payload
+        """
+        self.heartbeat_interval_ms = message["d"]["heartbeat_interval"]
+        asyncio.ensure_future(self.heartbeat())
+
+        handshake_identity = self.config["handshake_identity"]
+        asyncio.ensure_future(self.send_json({"op": Opcodes.IDENTIFY, "d": handshake_identity}))
+
+    async def heartbeat(self):
+        """
+        Sends a heartbeat payload every heartbeat interval.
+        """
+        asyncio.ensure_future(asyncio.sleep(self.heartbeat_interval_ms / 1000.0))
+        asyncio.ensure_future(self.send_json({"op": Opcodes.HEARTBEAT, "d": self.last_seq}))
+        self.event_loop.create_task(self.heartbeat())
+
+    async def gateway_handler(self):
+        """
+        Creates the websocket, receives messages and acts on them.
+        """
         async with websockets.connect("{}/?v={}&encoding={}".format(self.gateway_ws_url,
                                                                     self.config["gateway_api_version"],
                                                                     self.config["gateway_encoding"])) as websocket:
+            self.websocket = websocket
             while True:
-                message = await websocket.recv()
-                print(message)
+                message = asyncio.ensure_future(self.websocket.recv())
```

Comment:
This won't work since this won't wait for a message to be received before trying to process it. Make sure to always test your code after making major changes like this!